src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -1058,6 +1119,136 @@ impl Tree {
+        alloc_id: AllocId, // diagnostics
+        span: Span,        // diagnostics
+    ) -> InterpResult<'tcx> {
+        if let Some((access_range, access_kind, access_cause)) = access_range_and_kind {
+            for (loc_range, loc) in self.locations.iter_mut(access_range.start, access_range.size) {
+                // We can simply iterate over the entire graph in arbitrary order, as the
+                // relatedness of the access for each node can be calculated with only
+                // the tracking information attached to each node.
+                let mut stack = vec![self.root];
+                while let Some(id) = stack.pop() {
+                    let node = self.nodes.get_mut(id).unwrap();
+                    let mut entry = loc.perms.entry(id);
+                    let perm = entry.or_insert(node.default_location_state());
+                    // TODO narrowing based on protector
+                    let protected = global.borrow().protected_tags.contains_key(&node.tag);
+                    let mut entry = loc.wildcard_accesses.entry(id);
+                    let wildcard_access = entry.or_insert(Default::default());
+
+                    let Some(wildcard_relatedness) =
+                        wildcard_access.access_relatedness(access_kind)
+                    else {
+                        // there doenst exist a valid exposed reference for this access
+                        // to happen through
+                        // if this fails for one id, then it fails for all ids
+                        return Err(no_valid_exposed_references_error(
+                            alloc_id,
+                            loc_range.start,
+                            access_cause,
+                        ))
+                        .into();
+                    };
+                    if let Some(relatedness) = wildcard_relatedness.to_relatedness()
+                        && matches!(
+                            perm.skip_if_known_noop(access_kind, relatedness),
+                            ContinueTraversal::SkipSelfAndChildren
+                        )
+                    {
+                        continue;
+                    }
+
+                    // add children to stack
+                    stack.extend(node.children.iter().copied());
+                    let Some(relatedness) = wildcard_relatedness.to_relatedness() else {
+                        // if the access type is either, then we do not apply any transition to this node,
+                        // but we still update each child
+                        continue;
+                    };
+
+                    //update idempototent foreign access data
+                    if matches!(
+                        perm.skip_if_known_noop(access_kind, relatedness),
+                        ContinueTraversal::Recurse
+                    ) {
+                        perm.record_new_access(access_kind, relatedness);
+                    }
```

Comment:
I don't think this should be inside an `if`. This always comes together with `perm.perform_access`.